server/src/main/java/org/elasticsearch/discovery/single/SingleNodeDiscovery.java

```diff
@@ -101,7 +103,15 @@ public synchronized void startInitialJoin() {
         // apply a fresh cluster state just so that state recovery gets triggered by GatewayService
         // TODO: give discovery module control over GatewayService
         clusterState = ClusterState.builder(clusterState).build();
-        clusterApplier.onNewClusterState("single-node-start-initial-join", () -> clusterState, (source, e) -> {});
+        clusterApplier.onNewClusterState("single-node-start-initial-join", () -> clusterState, new ClusterApplyListener() {
```

Comment:
Possible improvement here: have a default (empty) `onSuccess` implementation so this can be a lambda.